### PR TITLE
Bug openemr fix 7746 export since delta fix

### LINKED ISSUE #7746: bug: Bulk export _since parameter is throwing error
As reported on Alan from Cushla on the community forum:
[alan-cushla](https://community.open-emr.org/u/alan-cushla) [forum bug report](https://community.open-emr.org/t/error-using-since-param-in-fhir-bulk-export/23564)

**Situation**
I have set up the fhir api, created the client assertion etc. and all is working well. I was hoping to use the '_since' parameter so that I can do a daily delta job. However, I'm getting an error. This is the url I'm using: `http://localhost/apis/default/fhir/$export?_since=2023-01-01T00:00:00Z`
and the error (from the error log) is: 
```
Uncaught TypeError: OpenEMR\\FHIR\\Export\\ExportJob::setResourceIncludeTime(): Argument #1 ($resourceIncludeTime) must be of type DateTime, string given, called in /var/www/localhost/htdocs/openemr/src/RestControllers/FHIR/Operations/FhirOperationExportRestController.php on line 130 and defined in /var/www/localhost/htdocs/openemr/src/FHIR/Export/ExportJob.php:216\nStack trace:\n#0 /var/www/localhost/htdocs/openemr/src/RestControllers/FHIR/Operations/FhirOperationExportRestController.php(130): OpenEMR\\FHIR\\Export\\ExportJob->setResourceIncludeTime()\n#1 /var/www/localhost/htdocs/openemr/_rest_routes.inc.php(13333): OpenEMR\\RestControllers\\FHIR\\Operations\\FhirOperationExportRestController->processExport()\n#2 /var/www/localhost/htdocs/openemr/src/Common/Http/HttpRestRouteHandler.php(103): {closure}()\n#3 /var/www/localhost/htdocs/openemr/apis/dispatch.php(389): OpenEMR\\Common\\Http\\HttpRestRouteHandler::dispatch()\n#4 {main}\n  thrown in /var/www/localhost/htdocs/openemr/src/FHIR/Export/ExportJob.php on line 216
```
It says the issue is in the `ExportJob.php` file but I think it actually originates in the `FhirOperationExportRestController.php` file in this line:
```
$since = $exportParams['_since'] ?? new \DateTime(date("Y-m-d H:i:s", 0)); // since epoch time
```
I think if the param is there, then a string is being using and it should be converted to a date first.
I know very little about php so I could be on the wrong track entirely here but I've tried a bunch of different formats, long, short, epoch time, etc. and I always get the same error.

**OpenEMR Version**
I'm using the flex-3.20 container.

**Browser**: 
I'm using: Edge

**Operating System**
I'm using: Linux and Windows

**Search**
Did you search the forum for similar questions? Yes.

**Logs**
```
Uncaught TypeError: OpenEMR\\FHIR\\Export\\ExportJob::setResourceIncludeTime(): Argument #1 ($resourceIncludeTime) must be of type DateTime, string given, called in /var/www/localhost/htdocs/openemr/src/RestControllers/FHIR/Operations/FhirOperationExportRestController.php on line 130 and defined in /var/www/localhost/htdocs/openemr/src/FHIR/Export/ExportJob.php:216\nStack trace:\n#0 /var/www/localhost/htdocs/openemr/src/RestControllers/FHIR/Operations/FhirOperationExportRestController.php(130): OpenEMR\\FHIR\\Export\\ExportJob->setResourceIncludeTime()\n#1 /var/www/localhost/htdocs/openemr/_rest_routes.inc.php(13333): OpenEMR\\RestControllers\\FHIR\\Operations\\FhirOperationExportRestController->processExport()\n#2 /var/www/localhost/htdocs/openemr/src/Common/Http/HttpRestRouteHandler.php(103): {closure}()\n#3 /var/www/localhost/htdocs/openemr/apis/dispatch.php(389): OpenEMR\\Common\\Http\\HttpRestRouteHandler::dispatch()\n#4 {main}\n  thrown in /var/www/localhost/htdocs/openemr/src/FHIR/Export/ExportJob.php on line 216
```

### PR DESCRIPTION
**Pull request by adunsulag**
This fixes the problems with the _since parameter in the export.  It was essentially being ignored and only grabbing data from the beginning of time for the entire system.

The main process flow is adding to the IFHIRExportableResourceService the getLastModifiedSearchField().  Every exportable resource implements this resource (can be nullable).  The _since FHIR instant date for the bulk export is passed to each exportable search field and all of the search queries will then auto-filter on the last modified date.  this is done internally by using the FHIR search modifier of 'ge'.  

So a _since parameter of ```2024-10-03T21:03:08+00:00``` is equivalent to calling a resource with the **_lastUpdated** search param of ```ge2024-10-03T21:03:08+00:00```

For example an AllergyIntolerance to an OpenEMR server running on local host would be
```
https://localhost:9300/apis/default/fhir/AllergyIntolerance?_lastUpdated=gt2024-09-03T21%3A03%3A08%2B00%3A00
```

Note that because its a DateTime filter for _lastUpdated you can do fuzzy matching and search modifiers such as greater than or equal to (ge) less than or equal to (se) greater than (gt), etc.  See the FHIR search spec for more details.

Had to add several database columns to handle last_updated pieces.  Also to avoid confusion, added date_created columns on elements that had no creation date tracking also.

Database changes were the following:
- users -> date_created, last_updated
- facility -> date_created, last_updated
- insurance_companies -> date_created, last_updated
- facility_user_ids -> date_created, last_updated (handles user / facility roles & specialties)
- openemr_postcalendar_categories -> pc_last_updated
- list_options -> last_updated
- form_clinical_notes -> last_updated
- form_vitals -> last_updated
- procedure_providers -> date_created, last_updated
- drugs -> date_created, last_updated
- patient_data -> last_updated

 Note this fixes #7746

### CODE REVIEW

**Comment by adunsulag** [comment]
Code is complete.  Added a number of database fields @bradymiller so probably want to pay close attention that.  Now that the PR is updated, will pull onto inferno test bed system and verify everything is still working.

File: sql/7_0_2-to-7_0_3_upgrade.sql [hunk 1]
```diff
@@ -134,4 +134,72 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
 #IfNotRow4D supported_external_dataloads load_type ICD10 load_source CMS load_release_date 2024-10-01 load_filename Zip File 3 2025 ICD-10-PCS Codes File.zip
 INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_release_date`, `load_filename`, `load_checksum`) VALUES
 ('ICD10', 'CMS', '2024-10-01', 'Zip File 3 2025 ICD-10-PCS Codes File.zip', 'a47ceb9a09fcc475fec19cee6526a335');
-#EndIf
\ No newline at end of file
+#EndIf
+
+#IfMissingColumn users date_created
+ALTER TABLE `users` ADD `date_created` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP;
```
Comment (bradymiller) [inline sql/7_0_2-to-7_0_3_upgrade.sql:140]: since this is a date_created, shouldn't have the ON UPDATE

File: sql/7_0_2-to-7_0_3_upgrade.sql [hunk 2]
```diff
@@ -134,4 +134,72 @@ INSERT INTO `supported_external_dataloads` (`load_type`, `load_source`, `load_re
...
+#EndIf
+
+#IfMissingColumn users last_updated
+ALTER TABLE `users` ADD `last_updated` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP;
+#EndIf
+
+#IfMissingColumn facility date_created
+ALTER TABLE `facility` ADD `date_created` DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP;
```
Comment (bradymiller) [inline sql/7_0_2-to-7_0_3_upgrade.sql:148]: remove ON UPDATE

File: sql/database.sql [hunk 3]
```diff
@@ -1485,6 +1485,8 @@ CREATE TABLE `drugs` (
   `drug_code` varchar(25) NULL,
   `consumable` tinyint(1) NOT NULL DEFAULT 0 COMMENT '1 = will not show on the fee sheet',
```
Comment (bradymiller) [inline sql/database.sql:1486]: upgrade scripts adds 17 last_updated/date_created items, but only 15 in the database.sql. So looks like database.sql is missing 2 :)

**Comment by bradymiller** [comment]
code review complete

File: sql/7_0_2-to-7_0_3_upgrade.sql (hunk 1 above)
Comment (adunsulag) [reply sql/7_0_2-to-7_0_3_upgrade.sql]: Thanks, fixed.

File: sql/7_0_2-to-7_0_3_upgrade.sql (hunk 2 above)
Comment (adunsulag) [reply sql/7_0_2-to-7_0_3_upgrade.sql]: Thanks, fixed.

File: sql/database.sql (hunk 3 above)
Comment (adunsulag) [reply sql/database.sql:1486]: Good catch! Haha!  Got it in.

**Comment by adunsulag** [comment]
Reminder to myself @adunsulag to bring in as a separate PR an update to the EHI documentation to cover the new columns and fields once this is brought in.

**Comment by adunsulag** [comment]
Need to fix the fresh database issue.  Verified Inferno is passing.  Need a future issue to improve performance w/ db indexes etc.

[OpenEMR-onc-report-2024-10-09.pdf](https://github.com/user-attachments/files/17306932/OpenEMR-onc-report-2024-10-09.pdf)

image

**Comment by adunsulag** [comment]
Going to bring this in now that inferno is passing.  We'll see if anyone hits any problems on demo testing but we can now bring this into patch.

**Comment by bradymiller** [comment]
dance-bear

**Pull request merged**